### ISSUE TITLE
Potential security concerns

### ISSUE THREAD

**Issue by 0x4E69676874466F78** [chaiNNer-org/chaiNNer-rs#16]
I'm really scared that you've started using this "programming language".
I looked at the 190 packages/crates used here.
There could be malware in any of these packages.
For example, from the world of npm: 
https://www.infosecurity-magazine.com/news/hundreds-malicious-packages-npm/
https://gist.github.com/MidSpike/f7ae3457420af78a54b38a31cc0c809c
In Rust, everything is based on faith. 
Rust doesn't solve the supply chain attack problem https://en.wikipedia.org/wiki/Supply_chain_attack
My advice is to put all dependencies inside the repository and update manually by auditing.
Also I strongly advise you not to use vscode in conjunction with rust-analyzer and not to run `cargo build` of any unverified code.
In case you didn't know:
https://github.com/eleijonmarck/do-not-compile-this-code
https://github.com/lucky/bad_actor_poc
https://rust-analyzer.github.io/manual.html#security
In Rust, macros evaluates on your machine and absolutely any code can be executed.
Also you use serde, there was such a scandal there recently: https://github.com/serde-rs/serde/issues/2538

Be safe 💕.

**Comment by RunDevelopment** [comment]
Let me start by saying that I think that the concerns you raised are valid. Arbitrary code execution being a feature of build scripts and proc macros is a real security concern, and one that the Rust language itself should address.

That being said, I am not particularly concerned about it in the context of this repo. This repo does not have many deps. Most of the deps I added are packages that commonly used all throughout the Rust ecosystem, so the likelihood of malware inside of them going undetected seems low (although popular packages are a particularly appealing target of course). I also locked the versions of our dependencies with a cargo lock file, so if a malicious version of some package were to be published, we would be unaffected (assuming that our current versions are unaffected, of course).

I also want to point out that npm and pip have the exact same issue as Rust, but on an even larger scale. npm and pip both support build scripts that are run on installation and execute arbitrary code. So while this in no way makes Rust any better (well, aside from pip due to the lack of lock files), the same issue is much more pronounced for npm and pip. On top of that, we have many more npm and pip dependencies than Rust dependencies (right now, and I do plan to keep it that way). 

While I do understand wanting to minimize the possible attack surface (by cutting down another package ecosystem), this will also come at a cost. We do not use Rust simply because we can. Rust solves real issues for us. The most important ones being 1) performance and 2) functionality.

1. We had a few nodes implemented in python that performed *very* badly and used Rust to solve these performance issues. E.g. the new Rust implementation of the dithering nodes is between 100x and 1000x faster.
2. Rust, despite being fairly young, already has a very powerful ecosystem around it. We use the functionality provided by that ecosystem to fill in the holes of Python and JS. E.g. we use Rust's regex engine to guarantee linear runtime in both Python and JS. We also use a rust crate to copy images to clipboard, since there exists no python library that can do it cross-platform.

Of course, C and C++ also have these features, but there are a lot of reasons why we went with Rust instead.

- Rust guarantees memory safety. While this might sound like a minor feature, this is huge. According to both Microsoft and Google, 70% of all security vulnerabilities in C and C++ code are due to memory management issues. Rust completely solves this.
- Cargo. While dependencies bring their own problems, they also solve other problems. The lack of standard dependency manager in C/C++ is a huge issue, and means that it is a pain to use libraries. This only gets worse when you need libraries that work across different platforms.
- Build & deploy. I am not very experienced in neither Rust nor C++ (having only worked on 2 or 3 non-trivial projects in both languages). The very fact that it took me like 15 minutes from knowing nothing about building pip wheels to cross-compiling across 8 architectures and 4 python versions should speak for itself.
- Productivity. The fact that Rust makes so many guarantees about correctness and safety means that I don't really have to worry about the same details and focus on the problem at hand. If I get the details wrong (memory, data race, invalid use of libraries), Rust will throw a compiler error my way, so I can immediately fix while I'm writing code.
- I just plain like Rust better. Rust is the most loved language for a reason, and my experience with C++ has been less than stellar.

So Rust is likely here to stay, because there are no good alternatives.

---

To address some of your individual points in more detail.

> There could be malware in any of these packages.

Yes. The trusting trust model of most package managers (cargo, npm, pip, whatever) is frankly not sufficient, but I unfortunately am not aware of any scalable solution for this.

That being said, the Rust community is keenly aware of this issue. E.g. there have been discussions around formalizing the chain of trust, and making that a feature of cargo, so you would know whose trust you're trusting (among other features this would enable, such as only allowing deps that have been verified by people you trust). That being said, such a system is still far from being implemented into cargo, unfortunately.

> My advice is to put all dependencies inside the repository and update manually by auditing.

Not a fan. While [cargo does support this](https://doc.rust-lang.org/cargo/commands/cargo-vendor.html), I do not see the necessity. We already use a cargo lock file, so we already have the benefit of having locked dependencies. As for review: transitive dependencies quickly blow up any dependency graph, so asking one developer (me) to review potentially thousands of lines of code to update a single dependency just doesn't scale. I'm also not confident that I would actually be able to detect potential malicious software better than the industry professionals that are already reviewing crates.

> Also I strongly advise you not to use vscode in conjunction with rust-analyzer and not to run cargo build of any unverified code.

I'm not sure how this would help. Suppose I turned off proc macros and build scripts rust-analyzer, I would still run the program to test it. Whether the trojan gets run by RA or me doing `cargo test` doesn't matter. There's simply no additional security gained here, but the loss in IDE features is significant.

> In Rust, macros evaluates on your machine and absolutely any code can be executed.

Yes, build scripts are a security concern, and one that Rust takes very seriously. [They recently removed a crate](https://blog.phylum.io/rust-malware-staged-on-crates-io/) that seems to have experimented with using `build.rs` to send personal info to a remote server. Scary stuff.

That being said, npm has the exact same issue. npm also supports and runs build scripts on install. Same for pip.

The only difference between npm, pip, etc and cargo is that there are actual plans to fix this issue for rust. Right now, there are 2 types of arbitrary code that rust executes on build: proc macros and build scripts. There is active movement towards using WASM to sandbox macro execution, which completely solves the security issue. Build scripts, by their very nature, are harder to sandbox, but there are also discussions about supporting WASM for them, such that we at least sandbox the ones that can be sandboxed.

In summary: I am less concerned about security risks emanating from cargo and crates.io than I am npm and pip. Rust has safety and security as one of its core community values, and a lot of work is being done to mitigate or even outright solve the security concerns you raised. So I can only see rust as a shining example in that regard, as I am not aware of any such movements/discussions around npm and pip, where it seems to have been just accepted. Of course, that might just be selection bias, since I read more about Rust than JS/Python.

**Comment by RunDevelopment** [comment]
Also, why "programming language" lol

**Comment by RunDevelopment** [comment]
> https://medium.com/packagr/using-pip-compile-to-manage-dependencies-in-your-python-packages-8451b21a949e
> https://stackoverflow.com/questions/66751657/what-does-pip-compile-do-what-is-its-use-how-do-i-maintain-the-contents-of-my

- `pip freeze`: Pro: deps are locked. Con: `requirements.txt` becomes unusable. Also, when I tested `pip freeze`, it seems to just output *all* python packages installed on my system, and not from the `requirements.txt` of my project (even with `-r requirements.txt`). Maybe I was doing it wrong, but this command seems to be not very useful for project management outside venvs.
- `pip-compile`: This is pretty cool. It fully solves the lock file issue. Why is that not part of pip... (Also, some smart scammers registered the domain `requirements.in`. So when you type `requirements.in` in your search bar to google it, you'll be met with a surprise :) )

Thanks for letting me know about `pip-compile`. I think I'll use that in the future.

> Maybe in the future you should try Mojo

I'm sorry, but I cannot see this a serious suggestion. At least judging from their docs, Mojo seems to be so young that major language features simply aren't implemented yet. E.g. Mojo seems to have an ownership and borrowing memory management model, but pretty much all mention of it are followed by ["TODO"](https://docs.modular.com/mojo/programming-manual.html#lifetimes).

So it would certainly be nice if Mojo became a fully-featured system programming language within the next few years, but I want to write fast code right now...

<details>
<summary>Tangent; my thoughts on Mojo's future</summary>

Also, this is a bit of a tangent, but I'll say it anyway: I do not think Mojo will make it big. Its main appeal seems to be: "we let you write python code, but it's fast." I just don't think that this (if they can achieve this goal) is enough to get Rust levels of adoption. I think this for multiple reasons. While they might be able to support all of Python's syntax, they cannot support all of Python's semantics (an obvious example would that python has a GC), so pure python devs will have a hard time learning Mojo (int types, lifetimes, memory management, etc.). It also puts huge design limitations on Mojo, because they have to mimic python semantics somehow, because otherwise Mojo would just be C with different syntax. They have chosen a very bumpy road, and I just cannot seem them being widely adopted at the end of it. I do see them becoming a language a few python developers will use for faster code, but I cannot see them be more than a very niche language.
A very bleak prediction, I know. I honestly wish for them to succeed, but I just don't think they'll make it.

</details>

> Rust doesn't solve these problems completely.

True. I should have been more specific. You're right that this statement does not apply to unsafe Rust.

Luckily, most crates contain little `unsafe`, if any at all.

> Moreover in Rust, even "unsafe" is just a syntactic construct. very simple example:
> 
> ```rust
> fn main() {
>   let unsafety = unsafe{||core::hint::unreachable_unchecked()};
>   unsafety(); // boom in safe context, no compiler warnings
>   print!("safety");
> }
> ```

You seem to have misunderstood the concept of `unsafe`. I say this not just because of this example code, but also because of your other points about unsafe code in Rust.

Unsafe Rust is an extension of the core (safe) language to allow for more low-level control where necessary. As the Rust compiler guarantees certain forms of safety, there will be situations where you (the programmer) cannot prove to the Rust compiler that your code is correct. This is simply a consequence of having a sound type/life time system (incompleteness theorem). The unsafe system allows you to say to the compiler "the language itself is not enough to prove that this code is correct, but trust me, it is." So the unsafe system exists because a strict type system + ownership + borrowing make it really difficult or even impossible to implement certain algorithms and data structures in Rust. (E.g. linked lists and graphs are (in)famously hard to implement in Rust.)

That's why unsafe Rust exists, but what does `unsafe` mean specifically. The `unsafe` keyword in Rust has an unchecked contract attached to it: you are responsible for upholding the guarantee of safe Rust. So if you use `unsafe`, and the code that you just told the compiler is okay is, in fact, not okay, then it's UB. And since safe Rust does not allow UB, this means that any `unsafe` black/function that can expose UB to safe Rust is incorrect. So an UB in safe Rust is a bug that can be traced down to some `unsafe` block. No ifs and buts. Consequently, unsafe Rust is supposed to be used to create safe wrappers around unsafe systems, and this is also how `unsafe` is (typically) used in practice.

Going back to your code example: As state in its documentation, calling `core::hint::unreachable_unchecked` is UB. So you broke the contract of `unsafe`, because you allowed this function to be called from safe Rust. As always, we can track down the source of UB to a single `unsafe`.

So does unsafe Rust compromise safety in Rust? Kinda. Bugs in unsafe code compromise the safety of safe Rust code. But bugs are just that: bugs. They will be fixed eventually. And since you only need to review `unsafe` code, those bugs should also be easier to find than in other languages (not just C and C++, e.g. I can have type casts everywhere in TS and invalidate type safety).

> Crates are full of unsafe and bindings to C.

Well, that's a claim :) But enough with my snarky remarks. Here's [the first study I found](https://cs.stanford.edu/~aozdemir/blog/unsafe-rust-syntax/) on that topic (It was literally the first link on Google and I didn't search any further, so there might be better studies out there). It says ~30% of crates and ~5% of code (measured in functions and blocks). So the amount of `unsafe` in crates is significant and should definitely be reduced, but I wouldn't call that "full of". Most crates don't contain `unsafe` at all.

> Rust doesn't even have its own standard library, it's [based on libc](https://news.ycombinator.com/item?id=9435804) (SAFETY).

Why does `std` wrapping libc not make it a standard library? It's not like `std` is just libc, there's a lot more to it. 

And about safety: it's not really an issue because `std` is a safe wrapper around libc, so it can't be a source of UB in safe Rust (absent bugs). But if you're concerned, there are some Rust purists that began reimplementing `std` in pure Rust. It's not fully functional yet, though. (Interestingly, a pure Rust implementation of `std` will also mean that `std` (or substantial parts of it) will be available on embedded devices, as libc currently prevents the use `std` there AFAIK.)

> You need to realize that old C++ != new C++.

My experience has been C++17, where I have written purely what most would likely consider modern C++ (at least at the time when C++17 was the newest version). Parts of one code base were written in C++03. So I'm well aware of the improvement C++ made since the "C with classes" days. I also keep an eye on the development of the language, so I at least heard of and seen major C++20 and 23 features.

Unfortunately, the issues with C++ that I raised (memory safety, lack of standard package manager, build systems) were not solved since C++17. I might as well add [null safety](https://godbolt.org/z/cha6M7M9T) and data races into the mix...

> "Rust is the most loved language for a reason" — it's fanboy-sectarian sentiment

Yet, it is also backed up by data. E.g. the SO developer survey, which puts Rust not only as the most loved/admired language for 8 years in a row now. Rust usage has also grown every year. So clearly, it's not just a few fanboys that like the language.

> Pardon my rudeness.

No worries, friend.

> No need to rebroadcast propaganda.

Me rn https://www.youtube.com/watch?v=khRJMiquAjA

**Comment by 0x4E69676874466F78** [comment]
> Going back to your code example
```rust
unsafe fn unsafety_fn() {
    core::hint::unreachable_unchecked()
}
fn main() {
  //let unsafety = unsafe{||core::hint::unreachable_unchecked()};
  //unsafety(); // boom in safe context, no compiler warnings
  unsafety_fn(); // error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
  print!("safety");
}
```
It should happen with the lambda but it doesn't. It's just another hole that's been forgotten about. 
I can show you other example without unreachable_unchecked().
Although this hole can still be fixed unlike [fake_static](https://github.com/meithecatte/fake-static).
Also PhantomData, mem::forget stories...

> but I wouldn't call that "full of".

The difficulty of expressing thoughts in English. There are "enough".
Anything that has to do with performance and flexibility is usually "unsafe".

> embedded devices

Rust in debug is so fat and clumsy that it can't even get the USB to start.
```
https://github.com/rust-embedded-community/usb-device/issues/80
https://github.com/stm32-rs/synopsys-usb-otg/issues/21
```
(I don't want to make mention)

> data races 

https://hirrolot.github.io/posts/rust-is-hard-or-the-misery-of-mainstream-programming.html#the-problem-with-rust

> So clearly, it's not just a few fanboys that like the language.

[Pic](https://github.com/chaiNNer-org/chaiNNer-rs/assets/4449851/765bd9f3-cddb-49b4-8b4a-6cdb355855dd)

Rust is not yet such a popular language so it's easy for him to gain at the expense of fanboys "most loved language"
In statistics, percentages and there are no absolute values. 
https://www.reddit.com/r/rustjerk/comments/16n10l9/rust_be_like/

Also here is a modified network marketing scheme. 
Rust has the foundation and exclusive control of the brand (there was a scandal recently about brand policy, which led to the creation of fork "crablang"), there is a system of grants. There are attempts to abusive terms from other languages such as UB, through which the language is sold for uninformed/clueless people.
Much of the community is loaded with propaganda and tales of language. Example: https://www.youtube.com/watch?v=0rJ94rbdteE (it's not a one-off channel).
It's long to explain and I don't have time, sorry.
With experience I hope you'll understand what I've been writing to you about here.

**Comment by RunDevelopment** [comment]
> It should happen with the lambda but it doesn't.

See [here](https://stackoverflow.com/questions/27746662/can-i-create-an-unsafe-closure). Closures are just anonymous types that implement a trait, so the call method itself cannot be `unsafe`. I guess the compiler just moves the `unsafe {}` inside the closure to make it work.

That's why it behaves like that, but I agree that this code communicates the wrong message. At the very least, there should a clippy lint to warn about this. So I suggested this as a lint: https://github.com/rust-lang/rust-clippy/issues/11657

> Although this hole can still be fixed unlike [fake_static](https://github.com/meithecatte/fake-static).

Yeah, that's a pretty bad bug in the compiler. But it's still that: a bug. It's hard to fix, so it has been open for a while, but it should get fixed eventually. So I have no reason to believe that the bug will exist forever.

> PhantomData

What's the issue with that? It's a necessary type for type safety. I know that many people don't like it, but I used it for, and thought it was a quite elegant solution to variance.

> mem::forget

True. I still don't know why they thought that leaking memory=safe is a good default behavior... 

Although there has been recent community discussion about fixing this by adding a trait that tells you whether a type can be leaked safely. So I have hope that this will be fixed some day, since the community seems to be interested in fixing it.

> There are "enough".

That's very subjective. As I pointed out, most crates don't even contain unsafe code. So what target threshold would consider to be sufficiently to not impact safety anymore?

Related question: How does this then impact the safety guarantees of languages such as JS, Python, Java, and C#, where packages can include Rust and C/C++ code? Does the horribleness of unsafe native code invalidate the safety guarantees of those languages as well? Or put another way: would you say that Python is memory safe? And how do C/C++/Rust bindings impact that?

Of course, problems in other languages don't make Rust any better, but I don't like the purist "if it's not 100%, then nothing can be valid" argument.

> Anything that has to do with performance and flexibility is usually "unsafe".

Not sure what you mean with flexibility, but the performance part isn't really true. While `unsafe` undoubtably enables clever tricks to avoid unnecessary work, `unsafe` is not "usually" (=I read as "often" or "commonly") necessary for great performance. 

E.g. this project has 4 public functions [with `unsafe` in them](https://github.com/chaiNNer-org/chaiNNer-rs/blob/245e17fdb171f6504aa409b821c0057a269cb645/crates/image-core/src/util.rs#L13). They are all unstable `std` functions (mostly because people still discuss the right names for these things...) that I copied to use it in stable Rust. All 4 functions just prevent unnecessary copying, which is a nice optimization. Notably, neither these functions nor any `unsafe` is used in the various image operations implemented by this project. And not because I didn't try to optimize by using `unsafe`. I tried to eliminate various bounds checks using `unsafe`, only to find out that the compiler had already optimized them away safely. 
So in this project, which primarily exists for performance reasons, there 4 unsafe function, all of which should get included into the stable standard library with time.

Some other examples (I just picked some popular crates that I used before and need to perform calculations quickly, but aren't unsafe due to some C bindings/sys calls (I think)): 
- [`image`](https://github.com/search?q=repo%3Aimage-rs%2Fimage+unsafe+language%3ARust&type=code&l=Rust): 6 `unsafe` in 31.9K LOC, so 1 `unsafe` per 5300 LOC.
- [`imge-png`](https://github.com/search?q=repo%3Aimage-rs%2Fimage+unsafe+language%3ARust&type=code&l=Rust): no `unsafe` in 8.5k LOC.
- [`regex`](https://github.com/search?q=repo%3Arust-lang%2Fregex+unsafe+language%3ARust+NOT+path%3A%2F%5Eregex-capi%5C%2Fsrc%5C%2F%2F&type=code&l=Rust): 68 `unsafe` in 150k LOC, so 1 `unsafe` per 2200 LOC.
- [`rand`](https://github.com/search?q=repo%3Arust-random%2Frand+unsafe+language%3ARust&type=code&l=Rust): 25 `unsafe` in 19.5k LOC, so 1 `unsafe` per 780 LOC.
- [`ahash`](https://github.com/search?q=repo%3Atkaitchuck%2FaHash+unsafe+NOT+path%3A%2F%5Esmhasher%5C%2Fahash-cbindings%5C%2Fsrc%5C%2F%2F&type=code): 11 `unsafe` in 4.3k LOC, so 1 `unsafe` per 390 LOC

All of these crates that need to deliver high performance, yet they use `unsafe` sparingly. So I would say that the claim that performance "usually" comes with `unsafe`, is just not true. 

Of course, if that much `unsafe` per LOC qualified as "usually", then writing C/C++ would also "usually" come with security vulnerabilities, as Google found [around 1 CVE per 1000 LOC in C/C++](https://security.googleblog.com/2022/12/memory-safe-languages-in-android-13.html#:~:text=Historical%20vulnerability%20density%20is%20greater%20than%201/kLOC).

> Rust in debug is so fat and clumsy that it can't even get the USB to start.

I can't really comment on that, since I have no experience with embedded. I just saw conference talks from people working in embedded that talked about their (positive) experience with Rust. Their devices just had a few kilobytes of ROM IIRC, so they definitely didn't use debug builds.

> > data races
> 
> https://hirrolot.github.io/posts/rust-is-hard-or-the-misery-of-mainstream-programming.html#the-problem-with-rust

Uhmmm, so? Async Rust certainly has its problems, but data races aren't one of them. I should also mention that one of the major problems listed in the article, `async fn` in traits, will come very soon. The PR for it recently got merged. Stabilization might still take a little though.

> Rust is not yet such a popular language

I'm just going link [this year's SO survey](https://survey.stackoverflow.co/2023/#technology-most-popular-technologies) again, which puts it at number 14. SO might not be a perfect measure, but the trend is certainly clear. I would argue that this amount of current usage + Rust's growth over the years qualifies the label "popular".

> In statistics, percentages and there are no absolute values.

Then how do you know that Rust is *not* a popular language? Sorry, but I just don't like these types of "you can't know anything with certainty" arguments. Error bars exist for a reason, and we can still get information out of noisy data. That information then needs to be handled carefully, of course, but we can still obtain valid information.

> There are attempts to abusive terms from other languages such as UB

What does "abuse" mean here? 

If you meant that "the Rust community subverts what the term UB means", then UB in Rust has pretty much the same meaning as in C/C++. Aside from maybe minor differences in their exact details, they mean the same thing. 

If you meant that "saying UB=bad slanders C/C++", then I don't know what to tell you. UB is a necessary evil to allow for optimization. Since triggering UB means that your program is incorrect, including UB everywhere in the standard library is a tradeoff between correctness and performance. Ideally for correctness, there would be as little UB as possible. UB being so widespread in C++ has been a criticism of the language since long before Rust came out. I mean, there is a reason why basically every other language has opted to pay the cost for the correctness to not have UB. E.g. Java has to check `this` for `null` before every method call (ignoring optimizations) to not have UB.

> No need to rebroadcast propaganda. 
> network marketing scheme
> loaded with propaganda

Let's take a step back, please. Marketing and propaganda are tools to further an agenda. So I would like to ask you: 1) who is instigating this campaign and 2) for what reason. I'm asking this, because I can only see what you presented here as conspiratorial thinking. To be clear, I don't mean this as an insult or anything like that. I mean that a lot of unlikely things would have to be true in order for your statements to be correct (as I interpreted them). Almost like some entities had conspired to make them so.

Let me explain: large companies (e.g. Google, MS, AWS), governments (e.g. US government, I believe the EU too), and famous open source projects (e.g. the Linux kernel) have all come out with (at the very least) positive sentiment towards Rust. Famous C-purist and C++-hater Linus Torvalds has accepted Rust into the Linux kernel. Do you think that he fell for marketing, got influenced by propaganda, or succumbed to peer-pressure from his brainwashed colleagues? Or maybe did he just look at Rust, and think that it's an okay language that would be useful for the kernel? Do you think that the security researchers that created the recommendation published by the US government got paid to recommend Rust or were blinded by hype around the language? Or maybe they didn't like that C++ programmers can't seem to stop putting trivial security vulnerabilities (e.g. buffer overflows) into their code risking national security, and determined that Rust would help prevent some of those vulnerabilities. This is why I called it "conspiratorial thinking". You present it as if all of these entities have either fallen for the propaganda, or are in on it and are purposefully furthering the agenda.

It seems to be that you completely disregard the possibility that many people might actually have valid reasons for thinking that Rust is a better tool than other languages. That many people could have independently come to the same conclusion.

Rust doesn't have to be perfect (and oh god it isn't), it just needs to a big enough improvement over other languages to justify the pain of learning and switching to it. And it seems that many people think that the pain is worth the benefit they get from Rust. So why wouldn't people want to use a tool that they think is better than other tools? Why wouldn't those people then recommend this tool to others? Why wouldn't they want to advertise the tool when it means that the tools gets even more powerful (e.g. more crates)?

All of the hype and favorable ratings of Rust that we've seen can easily be explained with:

1. Rust is actually a good tool that solves a number of problems better than other similar tools.
2. People recognized this, and started using and sharing the tool.

Why would this natural process be deserving of such negative words like "marketing scheme" and "propaganda"? This would only be fitting if the actual tool (Rust) was bad, but it seems like most people don't judge it to be so.

---

As for how one would find Rust: Here's my story (or what I can remember of it). I first learned of Rust in late 2021. When I heard of Rust and its promises regarding safety while delivering performance, I thought: "if it was that easy, C++ would have done it years ago". But I kept watching the conference talk, and they explained the explicit ownership and borrowing system, and it clicked for me. All these things, I had to keep in my head and hopefully not mess up in C++, explicitly spelled out and checked by the compiler. A nice solution, but honestly not enough to blow me away. The talk continued and they quickly introduced enums and traits, but quickly skipped over them, to continue to the main topic (can't remember what it was). That made me curious. Having a background in Java (which taught me to hate null) and TS (which taught me to love types), so enums and traits looked really interesting. So I watched more conference talks about Rust the next couple of days, and learned that Rust aligns with I want from a programming language.

- I'm someone that will gladly spend hours thinking about API design if it means that I can enforce the constraints of that API using the tools of the language (typically the type system). So Rusts strict type system is a perfect fit for me (although I would like to see more types like integer range type and anonymous union types (e.g. like `A | B` in TS)).
- I'm someone that really likes thinking of clever data structures to optimize algorithms. So Rust's low-level nature gives me the control I need while also giving me the safety of high-level languages.
- I'm someone that doesn't like the invisible control flow of exceptions, so I immediately liked Rust's approach to error handling.
- I'm someone that hates implicitly nullable references (think Java), so `Option<T>` (and its clever optimizations) are still of one of my favorite aspects of Rust.
- As soon as I learned what sum types can do in Rust, I missed them in every other language I used. I'm serious. I will often make types in TS that are just sum types. They are super useful, and I honstly wish that more languages would add decent support for them. (Sorry C++, `std::variant` sucks compared to Rust enums. Java sealed interfaces actually seem pretty cool.)
- I don't like OOP class inheritance (Java taught me well), and Rust doesn't have it.
- I like more functional-leaning programming and iterators (C# LINQ is awesome!), and guess what Rust is pretty good at.

Don't get me wrong. Rust has many problems, but Rust is also currently the language that best aligns with what I want from a programming language.

---

Sorry for writing so much. My spellchecker refusing to check that much text should have been a warning to me, but I stil wanted to say some more and pressed on...

**Comment by 0x4E69676874466F78** [comment]
I'll probably respond to your post, but definitely not this year.
